### PR TITLE
Additional map properties during map creation

### LINKED ISSUE #345: Additional properties for map creation
As stated in todays meeting, some more customization options are needed for map creation.
Additional properties:

- [ ] privacy settings
- [x] description
- [x] location

### PR DESCRIPTION
**Pull request by ThorbenStaufer**
Adds three new properties (private flag, description, location) to maps, that can be entered during map creation.
The location property uses PostGIS to save the actual latitude/longitude coordinates in the database.
Those coordinates can be selected by the user through a clickable world map in the map creation form.

Resolves #345.

## Basics

- [ ] Details of what you changed are in commit messages.
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildserver is happy.
- [ ] The PR is rebased with current master.

## Checklist

- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I fixed all affected decisions
- [ ] I added unit tests for my code
- [ ] I added code comments, logging, and assertions as appropriate
- [ ] I translated all strings visible to the user
- [ ] I mentioned [every code or binary](/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)
- [ ] I created left-over issues for things that are still to be done
- [ ] Code is conforming to [our Architecture](/doc/architecture)
- [ ] Code is conforming to [our Guidelines](/doc/guidelines)
      (exceptions are documented)
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)

## Review

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] Examples are well chosen and understandable

### CODE REVIEW

**Comment by markus2330** [comment]
jenkins build please

**Comment by lukashartl** [comment]
jenkins build please

File: backend/Cargo.toml [hunk 1]
```diff
@@ -46,8 +46,10 @@ dotenvy = "0.15.0"
 typeshare = "1.0.0"
 derive_more = "0.99.17"
 uuid = { version = "1.3.2", features = ["serde", "v4"] }
+postgis_diesel = "2.1.0"
 log = "0.4.17"
 env_logger = "0.10.0"
+postgis_diesel = "2.1.0"
```
Comment (kitzbergerg) [inline backend/Cargo.toml:52]: Added twice. Remove one of them. Also move it up to the `# Data` section together with the other diesel dependencies.

File: backend/src/model/dto.rs [hunk 2]
```diff
@@ -254,6 +266,18 @@ pub struct MapSearchParameters {
...
     /// The owner of the map.
     pub owner_id: Option<i32>,
+    /// Wheter or not the map is private.
+    pub is_private: Option<bool>,
+}
+
+/// Support struct for transmitting latitude/longitude coordinates.
+#[typeshare]
+#[derive(Debug, Serialize, Deserialize)]
+pub struct LatLng {
```
Comment (kitzbergerg) [inline backend/src/model/dto.rs:276]: Maybe rename this to Coordinates?

File: backend/src/model/dto.rs [hunk 3]
```diff
@@ -254,6 +266,18 @@ pub struct MapSearchParameters {
...
+    /// Latitude of the point.
+    pub lat: f64,
+    /// Longitude of the point.
+    pub lng: f64,
```
Comment (kitzbergerg) [inline backend/src/model/dto.rs:280]: I would personally prefer then to be written out (latitude/longitude). If you want the short form then rename lng to long (that is the correct abbreviation according to wikipedia)

File: backend/src/model/dto/new_map_impl.rs [hunk 4]
```diff
@@ -17,6 +19,16 @@ impl From<NewMapDto> for NewMap {
...
+            is_private: new_map.is_private,
+            description: new_map.description,
+            location: match new_map.location {
+                Some(latlng) => Some(Point {
+                    x: latlng.lng,
+                    y: latlng.lat,
+                    srid: Some(4326),
+                }),
+                None => None,
+            },
```
Comment (kitzbergerg) [inline backend/src/model/dto/new_map_impl.rs:31]: Same as above.

File: backend/src/model/dto.rs [hunk 5]
```diff
@@ -254,6 +266,18 @@ pub struct MapSearchParameters {
     pub is_inactive: Option<bool>,
     /// The owner of the map.
     pub owner_id: Option<i32>,
+    /// Wheter or not the map is private.
```
Comment (kitzbergerg) [inline backend/src/model/dto.rs:269]:
```suggestion
    /// Whether or not the map is private.
```

File: backend/src/model/dto/map_impl.rs [hunk 6]
```diff
@@ -18,6 +18,15 @@ impl From<Map> for MapDto {
...
             owner_id: map.owner_id,
+            is_private: map.is_private,
+            description: map.description,
+            location: match map.location {
+                Some(latlng) => Some(LatLng {
+                    lat: latlng.y,
+                    lng: latlng.x,
+                }),
+                None => None,
+            },
```
Comment (kitzbergerg) [inline backend/src/model/dto/map_impl.rs:29]:
```suggestion
            location: map.location.map(|latlng| LatLng {
                    lat: latlng.y,
                    lng: latlng.x,
                }),
```

**Review by kitzbergerg — CHANGES_REQUESTED** [review]

**Comment by kitzbergerg** [comment]
> jenkins build please

It won't build until the `Cargo.toml` is fixed.

File: backend/Cargo.toml (hunk 1 above)
Comment (ThorbenStaufer) [reply backend/Cargo.toml:52]: I don't even know how that happened. I guess it slipped through during the merge.

File: backend/src/model/dto.rs [hunk 7]
```diff
@@ -218,6 +218,12 @@ pub struct MapDto {
     pub harvested: i16,
     /// The id of the owner of the map.
     pub owner_id: i32,
+    /// A flag indicating if this map is private or not.
+    pub is_private: Option<bool>,
```
Comment (markus2330) [inline backend/src/model/dto.rs:222]: We have 3 possible privacy settings, so it must be enum, see doc/usecases/current/map_creation.md

Why is it optional?

File: frontend/src/config/i18n/de/maps.json [hunk 8]
```diff
@@ -12,6 +12,9 @@
         "cancel_button": "Abbrechen",
         "submit_button": "Erstellen",
         "success": "Karte wurde erfolgreich erstellt!",
-        "failure": "Etwas ist schief gelaufen, bitte versuchen Sie es später erneut!"
+        "failure": "Etwas ist schief gelaufen, bitte versuchen Sie es später erneut!",
+        "private_label": "privat",
+        "description_placeholer": "Beschreibung",
+        "location_selected": "Dein gewählter Standort!"
```
Comment (markus2330) [inline frontend/src/config/i18n/de/maps.json:18]:
```suggestion
        "location_selected": "Dein gewählter Standort"
```

**Review by markus2330 — CHANGES_REQUESTED** [review]
Thanks for the PR! (Didn't test yet.)

File: backend/src/model/dto/new_map_impl.rs [hunk 9]
```diff
@@ -17,6 +19,16 @@ impl From<NewMapDto> for NewMap {
...
+            privacy: new_map.privacy,
+            description: new_map.description,
+            location: match new_map.location {
+                Some(latlng) => Some(Point {
+                    x: latlng.longitude,
+                    y: latlng.latitude,
+                    srid: Some(4326),
+                }),
+                None => None,
+            },
```
Comment (kitzbergerg) [inline backend/src/model/dto/new_map_impl.rs:31]:
```suggestion
            location: map.location.map(|latlng| Point {
                    x: latlng.longitude,
                    y: latlng.latitude,
                    srid: Some(4326),
            }),
```

If you use `From` and `Into` traits use the suggestion described in the other file.

File: backend/src/model/dto/new_map_impl.rs [hunk 10]
```diff
@@ -17,6 +19,16 @@ impl From<NewMapDto> for NewMap {
             visits: new_map.visits,
             harvested: new_map.harvested,
             owner_id: new_map.owner_id,
+            privacy: new_map.privacy,
+            description: new_map.description,
+            location: match new_map.location {
+                Some(latlng) => Some(Point {
+                    x: latlng.longitude,
+                    y: latlng.latitude,
+                    srid: Some(4326),
```
Comment (kitzbergerg) [inline backend/src/model/dto/new_map_impl.rs:28]: Why is there a hardcoded values? Does this mean something? If so please add a short comment about that.

File: backend/src/model/dto.rs [hunk 11]
```diff
@@ -254,6 +266,18 @@ pub struct MapSearchParameters {
...
     /// The owner of the map.
     pub owner_id: Option<i32>,
+    /// Whether or not the map is private.
+    pub privacy: Option<PrivacyOptions>,
+}
+
+/// Support struct for transmitting latitude/longitude coordinates.
+#[typeshare]
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Coordinates {
```
Comment (kitzbergerg) [inline backend/src/model/dto.rs:276]: You could try to implement `From<Point>` and `Into<Point>` for Coordinates. That way you can replace
```rust
            location: map.location.map(|latlng| Coordinates {
                latitude: latlng.y,
                longitude: latlng.x,
            }),
```
with
```rust
            location: map.location.map(From::from),
```

Same for
```rust
            location: match new_map.location {
                Some(latlng) => Some(Point {
                    x: latlng.longitude,
                    y: latlng.latitude,
                    srid: Some(4326),
                }),
                None => None,
            },
```
could be 
```rust
            location: new_map.location.map(Into::into)
```

**Review by kitzbergerg — COMMENTED** [review]
Just a few small code style things.

**Review by markus2330 — CHANGES_REQUESTED** [review]
The frontend does not load for me, I get the error:

`Uncaught (in promise) SyntaxError: import not found: PrivacyOptions ` MapCreateForm.tsx:2:20)

**Comment by ThorbenStaufer** [comment]
>The frontend does not load for me, I get the error:
>
>Uncaught (in promise) SyntaxError: import not found: PrivacyOptions MapCreateForm.tsx:2:20

It works for me. It seems you have forgotten to let typeshare generate the type.
You have to run `npm run generate-api-types`.

File: backend/src/model/dto.rs [hunk 12]
```diff
@@ -218,6 +218,12 @@ pub struct MapDto {
     pub harvested: i16,
     /// The id of the owner of the map.
     pub owner_id: i32,
+    /// A flag indicating if this map is private or not.
+    pub privacy: PrivacyOptions,
+    /// The description of the map.
+    pub description: Option<String>,
+    /// The location of the map as a latitude/longitude point.
+    pub location: Option<Coordinates>,
```
[the remark this answers is not in the document]
Comment (ThorbenStaufer) [reply backend/src/model/dto.rs:260]: Yes, location data can successfully be selected in the creation form and will be saved in the database.
A quick look with pgAdmin also confirms in the in-app Geo Map that the coordinates were correctly received.

**Comment by markus2330** [comment]
CI complains now:

```

src/features/maps/components/MapCreateModal.tsx(44,7): error TS2322: Type '{ name: string; creation_date: string; is_inactive: false; zoom_factor: number; honors: number; visits: number; harvested: number; owner_id: number; is_private: boolean; description: string; }' is not assignable to type 'NewMapDto'.

  Object literal may only specify known properties, and 'is_private' does not exist in type 'NewMapDto'.

```

File: frontend/src/config/i18n/de/maps.json [hunk 13]
```diff
@@ -12,6 +12,14 @@
...
-        "failure": "Etwas ist schief gelaufen, bitte versuchen Sie es später erneut!"
+        "failure": "Etwas ist schief gelaufen, bitte versuchen Sie es später erneut!",
+        "description_placeholer": "Beschreibung",
+        "location_selected": "Dein gewählter Standort.",
+        "private": "privat",
+        "private_info": "Die Karte ist privat und nur für dich sichtbar",
+        "protected": "geschützt",
+        "protected_info": "Die Karte ist geschützt und nur für dich und andere PermaplanT Mitglieder sichtbar",
+        "public": "öffentlich",
+        "public_info": "Die Karte ist öffentlich und für jeden sichtbar"
```
Comment (markus2330) [inline frontend/src/config/i18n/de/maps.json:23]:
```suggestion
        "public_info": "Die Karte ist öffentlich und für alle sichtbar"
```

**Review by markus2330 — APPROVED** [review]
I can create maps (it works nicely) but I don't always see the maps that were created immediately? Only after some time it refreshes and shows the created map. (Maybe it was only for the first map like this?)

More important: there is also no way to cllick on the maps. Slowly we should get rid of `/map` and instead have the functionality in `/maps/<id>`.

**Comment by ThorbenStaufer** [comment]
> Only after some time it refreshes and shows the created map.

I also noticed this but I honestly don't understand why it happens.
I used a lot of the code snippets from the seeds overview and I don't see any other reasons why the view is lagging behind.
But as you said, this phenomenon only happens to the first post request and subsequent creations show almost instantly, which only adds further questions.

> there is also no way to cllick on the maps.

I already opened an issue for this with #390. It was out of scope for this PR, as it is only concerned with the creation process itself, but it would be one of the next logical steps to tackle.

**Pull request merged**